File: backend/internal/api/response/response.go
package response

import (
	"encoding/json"
	"net/http"
)

type ErrorCode string

const (
	ErrorCodeValidation   ErrorCode = "VALIDATION_ERROR"
	ErrorCodeUnauthorized ErrorCode = "UNAUTHORIZED"
	ErrorCodeForbidden    ErrorCode = "FORBIDDEN"
	ErrorCodeNotFound     ErrorCode = "NOT_FOUND"
	ErrorCodeConflict     ErrorCode = "CONFLICT"
	ErrorCodeInternal     ErrorCode = "INTERNAL_ERROR"
)

type ErrorPayload struct {
	Code    ErrorCode   `json:"code"`
	Message string      `json:"message,omitempty"`
	Details interface{} `json:"details,omitempty"`
}

type SuccessBody struct {
	Success bool        `json:"success"`
	Data    interface{} `json:"data"`
}

type ErrorBody struct {
	Success bool         `json:"success"`
	Error   ErrorPayload `json:"error"`
}

func WriteSuccess(w http.ResponseWriter, status int, data interface{}) {
	writeJSON(w, status, SuccessBody{
		Success: true,
		Data:    data,
	})
}

func WriteError(
	w http.ResponseWriter,
	status int,
	code ErrorCode,
	message string,
	details interface{},
) {
	writeJSON(w, status, ErrorBody{
		Success: false,
		Error: ErrorPayload{
			Code:    code,
			Message: message,
			Details: details,
		},
	})
}

func WriteMethodNotAllowed(w http.ResponseWriter) {
	WriteError(
		w,
		http.StatusMethodNotAllowed,
		ErrorCodeValidation,
		"Method not allowed",
		nil,
	)
}

func writeJSON(w http.ResponseWriter, status int, payload interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)

	if err := json.NewEncoder(w).Encode(payload); err != nil {
		http.Error(
			w,
			`{"success":false,"error":{"code":"INTERNAL_ERROR","message":"Failed to encode response"}}`,
			http.StatusInternalServerError,
		)
	}
}

File: src/lib/rest-api/server/response.ts
import { NextResponse } from 'next/server'

export const ApiErrorCode = {
  VALIDATION_ERROR: 'VALIDATION_ERROR',
  UNAUTHORIZED: 'UNAUTHORIZED',
  FORBIDDEN: 'FORBIDDEN',
  NOT_FOUND: 'NOT_FOUND',
  CONFLICT: 'CONFLICT',
  INTERNAL_ERROR: 'INTERNAL_ERROR',
} as const

export type ApiErrorCode = (typeof ApiErrorCode)[keyof typeof ApiErrorCode]

export interface ApiErrorPayload {
  code: ApiErrorCode
  message?: string
  details?: unknown
}

export interface ApiResponseSuccess<T> {
  success: true
  data: T
}

export interface ApiResponseError {
  success: false
  error: ApiErrorPayload
}

export type ApiResponse<T> = ApiResponseSuccess<T> | ApiResponseError

type ApiErrorInit = {
  code: ApiErrorCode
  message?: string
  details?: unknown
  status?: number
}

export const buildApiSuccess = <T>(
  data: T,
  init?: { status?: number },
): NextResponse<ApiResponse<T>> => {
  return NextResponse.json(
    {
      success: true,
      data,
    },
    { status: init?.status ?? 200 },
  )
}

export const buildApiError = (
  payload: ApiErrorInit,
): NextResponse<ApiResponseError> => {
  return NextResponse.json(
    {
      success: false,
      error: {
        code: payload.code,
        message: payload.message,
        details: payload.details,
      },
    },
    { status: payload.status ?? 400 },
  )
}

export const isApiSuccess = <T>(
  value: ApiResponse<T>,
): value is ApiResponseSuccess<T> => value.success
